### PR TITLE
add mongodb writer to writer plugin manager

### LINKED ISSUE #37: MongoDB Writer is missing in WriterPluginManager

### PR DESCRIPTION
**Pull request by prolic**
resolves https://github.com/zendframework/zend-log/issues/37

### CODE REVIEW

File: test/LoggerAbstractServiceFactoryTest.php [hunk 1]
```diff
@@ -146,6 +147,53 @@ public function testRetrievesDatabaseServiceFromServiceManagerWhenEncounteringDb
         $this->assertAttributeSame($db, 'db', $writer);
     }
 
+    public function testRetrievesMongoDBServiceFromServiceManagerWhenEncounteringMongoDbWriter()
+    {
+        $mongoClient = $this->getMockBuilder('MongoClient')
```
Comment (weierophinney) [inline test/LoggerAbstractServiceFactoryTest.php:162]: This test should check for existence of the mongodb extension prior, and mark the test skipped if it is not. I can make that change on merge.

**Comment by weierophinney** [comment]
Ouch... there's another problem.

Evidently, Travis is installing ext/mongodb, vs ext/mongo, with current builds. ext/mongodb _does not ship the `MongoClient` or `Mongo` classes_ that were present in the various `1.*` builds. As such, tests currently do not run.

I'm not quite sure how to resolve this, but I'm a bit reluctant to expose the mongo writer unless we know it also works with ext/mongodb. Thoughts?

**Comment by prolic** [comment]
I'd say let's make two separate writers: 1) mongo 2) mongodb for ext-mongo and ext-mongodb.

**Comment by prolic** [comment]
About travis: I can install both extensions for travis, this is no big deal.

**Comment by sandrokeil** [comment]
I guess we run into the same problems as mentioned in the [zend-session component](https://github.com/zendframework/zend-session/pull/34#issuecomment-203146816). We should use an extra repo (e.g. zend-log-mongodb) for specific writer, which relies on dependencies.

**Comment by prolic** [comment]
@weierophinney If you want me to change this to separate repos, please create those repos for me so I can make the split. Otherwise it should be ready to merge.

**Comment by samsonasik** [comment]
ping @weierophinney

File: src/Writer/Mongo.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    protected $saveOptions;
+
+    /**
+     * Constructor
+     *
+     * @param MongoC|MongoClient|array|Traversable $mongo
+     * @param string $database
+     * @param string $collection
+     * @param array $saveOptions
+     * @throws Exception\InvalidArgumentException
```
Comment (samsonasik) [inline src/Writer/Mongo.php:49]: add `@throws Exception\ExtensionNotLoadedException`

Comment (prolic) [inline src/Writer/Mongo.php:49]: done.

File: test/Writer/MongoTest.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    public function testWriteConvertsDateTimeToMongoDate()
+    {
+        $date = new DateTime();
+        $event = ['timestamp'=> $date];
+
+        $this->mongoCollection->expects($this->once())
+            ->method('save')
+            ->with($this->contains(new MongoDate($date->getTimestamp()), false));
```
Comment (samsonasik) [inline test/Writer/MongoTest.php:91]: `MongoDB\BSON\UTCDateTime` ?

Comment (prolic) [inline test/Writer/MongoTest.php:91]: No, this is MongoTest, not MongoDBTest

File: src/Writer/Mongo.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        if ($mongo instanceof Traversable) {
+            // Configuration may be multi-dimensional due to save options
+            $mongo = ArrayUtils::iteratorToArray($mongo);
+        }
+        if (is_array($mongo)) {
+            parent::__construct($mongo);
+            $saveOptions = isset($mongo['save_options']) ? $mongo['save_options'] : [];
+            $collection  = isset($mongo['collection']) ? $mongo['collection'] : null;
+            $database    = isset($mongo['database']) ? $mongo['database'] : null;
+            $mongo       = isset($mongo['mongo']) ? $mongo['mongo'] : null;
```
Comment (weierophinney) [inline src/Writer/Mongo.php:67]: Wish we were on PHP 7 as the minimum version, so we could use `??` here. :frowning:

**Comment by weierophinney** [comment]
Merged to develop for release with 2.9.0.

**Pull request closed without merging**